crates/engine/tree/src/tree/config.rs

```diff
@@ -0,0 +1,103 @@
+//! Engine tree configuration.
+
+/// The configuration of the engine tree.
+#[derive(Debug)]
+pub struct TreeConfig {
+    /// Maximum number of blocks to be kept only in memory without triggering persistence.
+    persistence_threshold: u64,
+    /// Number of pending blocks that cannot be executed due to missing parent and
+    /// are kept in cache.
+    block_buffer_limit: u32,
+    /// Number of invalid headers to keep in cache.
+    max_invalid_header_cache_length: u32,
+}
+
+impl Default for TreeConfig {
+    fn default() -> Self {
+        Self {
+            persistence_threshold: 256,
+            block_buffer_limit: 256,
+            max_invalid_header_cache_length: 256,
+        }
+    }
+}
+
+impl TreeConfig {
+    /// Create engine tree configuration.
+    pub const fn new(
+        persistence_threshold: u64,
+        block_buffer_limit: u32,
+        max_invalid_header_cache_length: u32,
+    ) -> Self {
+        Self { persistence_threshold, block_buffer_limit, max_invalid_header_cache_length }
+    }
+
+    /// Return the persistence threshold.
+    pub const fn persistence_threshold(&self) -> u64 {
+        self.persistence_threshold
+    }
+
+    /// Return the block buffer limit.
+    pub const fn block_buffer_limit(&self) -> u32 {
+        self.block_buffer_limit
+    }
+
+    /// Return the maximum invalid cache header length.
+    pub const fn max_invalid_header_cache_length(&self) -> u32 {
+        self.max_invalid_header_cache_length
+    }
+
+    /// Return a builder for this type.
+    pub fn builder() -> TreeConfigBuilder {
+        TreeConfigBuilder::default()
+    }
+}
+
+/// Engine tree configuration builder.
+#[derive(Debug, Default)]
+pub struct TreeConfigBuilder {
+    persistence_threshold: Option<u64>,
+    block_buffer_limit: Option<u32>,
+    max_invalid_header_cache_length: Option<u32>,
+}
```

Comment:
I think the builder type is redundant here if we move the with setters to Treeconfig directly so you can configure it like `TreeConfig::default().with_().with()`

Reply:
makes sense, done